Get rid of `rustc_query_description!`

Noratrieb opened the pull request:
**I am not entirely sure whether this is an improvement and would like to get your feedback on it.**

Helps with #96524.

Queries can provide an arbitrary expression for their description and their caching behavior. Before, these expressions where stored in a `rustc_query_description` macro emitted by the `rustc_queries` macro, and then used in `rustc_query_impl` to fill out the methods for the `QueryDescription` trait.

Instead, we now emit two new modules from `rustc_queries` containing the functions with the expressions. `rustc_query_impl` calls these functions now instead of invoking the macro.

Since we are now defining some of the functions in `rustc_middle::query`, we now need all the imports for the key types mthere as well.

r? @cjgillot

Noratrieb committed:
Remove unsued variable in query macro

Noratrieb committed:
Get rid of `rustc_query_description!`

Queries can provide an arbitrary expression for their description and
their caching behavior. Before, these expressions where stored in a
`rustc_query_description` macro emitted by the `rustc_queries` macro,
and then used in `rustc_query_impl` to fill out the methods for the
`QueryDescription` trait.

Instead, we now emit two new modules from `rustc_queries` containing the
functions with the expressions. `rustc_query_impl` calls these functions
now instead of invoking the macro.

Since we are now defining some of the functions in
`rustc_middle::query`, we now need all the imports for the key types
there as well.

jyn514 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 6042ea72275fb466660bbdde8777dc7d6540e84b with merge 755919a27fb7a6a1ee85bc2810a3ec36c41099c6...
<!-- homu: {"type":"TryBuildStarted","head_sha":"6042ea72275fb466660bbdde8777dc7d6540e84b","merge_sha":"755919a27fb7a6a1ee85bc2810a3ec36c41099c6"} -->

cjgillot commented on compiler/rustc_query_impl/src/plumbing.rs:
```diff
@@ -466,7 +466,14 @@ macro_rules! define_queries {
 
         impl<'tcx> QueryDescription<QueryCtxt<'tcx>> for queries::$name<'tcx> {
-            rustc_query_description! { $name }
+            fn describe(tcx: QueryCtxt<'tcx>, key: Self::Key) -> String {
```
We never call `describe` through the trait. Could you remove it from the trait and make access more direct?

cjgillot commented on compiler/rustc_query_impl/src/plumbing.rs:
```diff
@@ -466,7 +466,14 @@ macro_rules! define_queries {
+            }
+
+            #[inline]
+            fn cache_on_disk(tcx: TyCtxt<'tcx>, key: &Self::Key) -> bool {
```
`cache_on_disk` is typically very small and never used in `rustc_query_system`.  I'd like to remove it from the trait too, if possible.  It's used in `try_load_from_on_disk_cache`, so it's a bit more complex as a refactor.

Nilstrieb replied:
Doing this in a not-cursed way (I have several cursed ideas involving conjuring up generic fn defs using `transmute(())`) requires a few more refactors than I'd like to do here, so I'd prefer doing that in a follow-up PR.

cjgillot commented on compiler/rustc_macros/src/query.rs:
```diff
@@ -237,27 +237,32 @@ fn doc_comment_from_desc(list: &Punctuated<Expr, token::Comma>) -> Result<Attrib
+            #[allow(unused_variables, unused_braces, rustc::pass_by_value)]
             #[inline]
-            fn cache_on_disk(#tcx: TyCtxt<'tcx>, #key: &Self::Key) -> bool {
+            pub fn #name<'tcx>(#tcx: TyCtxt<'tcx>, #key: &#arg) -> bool {
```
Could you use `query_keys::#name<'tcx>` instead of `#arg`?  This will allow us to avoid requiring `key_prelude`.

cjgillot commented on compiler/rustc_macros/src/query.rs:
```diff
@@ -237,27 +237,32 @@ fn doc_comment_from_desc(list: &Punctuated<Expr, token::Comma>) -> Result<Attrib
+            #[allow(rustc::pass_by_value)]
             #[inline]
-            fn cache_on_disk(_: TyCtxt<'tcx>, _: &Self::Key) -> bool {
+            pub fn #name<'tcx>(_: TyCtxt<'tcx>, _: &#arg) -> bool {
```
Likewise.

cjgillot reviewed (commented):
Thanks @Nilstrieb. I think this goes in the right direction.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3221650119/jobs/5271086135)
Build commit: 755919a27fb7a6a1ee85bc2810a3ec36c41099c6 (`755919a27fb7a6a1ee85bc2810a3ec36c41099c6`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3221650119/jobs/5271086135"},"merge_sha":"755919a27fb7a6a1ee85bc2810a3ec36c41099c6"} -->

rust-timer commented:
Queued 755919a27fb7a6a1ee85bc2810a3ec36c41099c6 with parent 0265a3e93bf1b89d97cae113ed214954d5c35e22, future [comparison URL](https://perf.rust-lang.org/compare.html?start=0265a3e93bf1b89d97cae113ed214954d5c35e22&end=755919a27fb7a6a1ee85bc2810a3ec36c41099c6).

rust-timer commented:
Finished benchmarking commit (755919a27fb7a6a1ee85bc2810a3ec36c41099c6): [comparison URL](https://perf.rust-lang.org/compare.html?start=0265a3e93bf1b89d97cae113ed214954d5c35e22&end=755919a27fb7a6a1ee85bc2810a3ec36c41099c6&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=0265a3e93bf1b89d97cae113ed214954d5c35e22&end=755919a27fb7a6a1ee85bc2810a3ec36c41099c6&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | 0.7%     | [0.2%, 0.9%] | 7         |
| Regressions ❌ <br /> (secondary) | 1.2%     | [1.0%, 1.3%] | 6         |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -0.8%    | [-1.1%, -0.4%] | 4         |
| All ❌✅ (primary) | 0.7%     | [0.2%, 0.9%] | 7         |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=0265a3e93bf1b89d97cae113ed214954d5c35e22&end=755919a27fb7a6a1ee85bc2810a3ec36c41099c6&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | -        | -     | 0         |
| Regressions ❌ <br /> (secondary) | -        | -     | 0         |
| Improvements ✅ <br /> (primary) | -4.1%    | [-4.1%, -4.1%] | 1         |
| Improvements ✅ <br /> (secondary) | -2.4%    | [-2.4%, -2.4%] | 1         |
| All ❌✅ (primary) | -4.1%    | [-4.1%, -4.1%] | 1         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=0265a3e93bf1b89d97cae113ed214954d5c35e22&end=755919a27fb7a6a1ee85bc2810a3ec36c41099c6&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | 2.2%     | [2.2%, 2.2%] | 1         |
| Regressions ❌ <br /> (secondary) | 3.7%     | [3.4%, 4.0%] | 3         |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -3.0%    | [-3.0%, -3.0%] | 1         |
| All ❌✅ (primary) | 2.2%     | [2.2%, 2.2%] | 1         |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

Noratrieb committed:
Remove the `describe` method from the `QueryDescription` trait

It was called directly already, but now it's even more useless since it
just forwards to the free function. Call it directly.

Nilstrieb commented:
I implemented most of the fixes, but I haven't investigated the perf regressions for diesel. Maybe another perf run will make them magically go away after my fixes? 🙂

jyn514 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 4ae32c8cca22265c5c3ea8c2b751a251f53f97ed with merge cce71dc4adbb716bd87c9a5208f331d6cbfc7382...
<!-- homu: {"type":"TryBuildStarted","head_sha":"4ae32c8cca22265c5c3ea8c2b751a251f53f97ed","merge_sha":"cce71dc4adbb716bd87c9a5208f331d6cbfc7382"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3245335000/jobs/5324426427)
Build commit: cce71dc4adbb716bd87c9a5208f331d6cbfc7382 (`cce71dc4adbb716bd87c9a5208f331d6cbfc7382`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3245335000/jobs/5324426427"},"merge_sha":"cce71dc4adbb716bd87c9a5208f331d6cbfc7382"} -->

rust-timer commented:
Queued cce71dc4adbb716bd87c9a5208f331d6cbfc7382 with parent 6b3ede3f7bc502eba7bbd202b4b9312d812adcd7, future [comparison URL](https://perf.rust-lang.org/compare.html?start=6b3ede3f7bc502eba7bbd202b4b9312d812adcd7&end=cce71dc4adbb716bd87c9a5208f331d6cbfc7382).

rust-timer commented:
Finished benchmarking commit (cce71dc4adbb716bd87c9a5208f331d6cbfc7382): [comparison URL](https://perf.rust-lang.org/compare.html?start=6b3ede3f7bc502eba7bbd202b4b9312d812adcd7&end=cce71dc4adbb716bd87c9a5208f331d6cbfc7382&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=6b3ede3f7bc502eba7bbd202b4b9312d812adcd7&end=cce71dc4adbb716bd87c9a5208f331d6cbfc7382&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | -        | -     | 0         |
| Regressions ❌ <br /> (secondary) | 1.1%     | [1.1%, 1.1%] | 3         |
| Improvements ✅ <br /> (primary) | -0.3%    | [-0.4%, -0.2%] | 14        |
| Improvements ✅ <br /> (secondary) | -0.7%    | [-1.2%, -0.3%] | 8         |
| All ❌✅ (primary) | -0.3%    | [-0.4%, -0.2%] | 14        |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=6b3ede3f7bc502eba7bbd202b4b9312d812adcd7&end=cce71dc4adbb716bd87c9a5208f331d6cbfc7382&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | -        | -     | 0         |
| Regressions ❌ <br /> (secondary) | -        | -     | 0         |
| Improvements ✅ <br /> (primary) | -2.7%    | [-2.7%, -2.7%] | 1         |
| Improvements ✅ <br /> (secondary) | -        | -     | 0         |
| All ❌✅ (primary) | -2.7%    | [-2.7%, -2.7%] | 1         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=6b3ede3f7bc502eba7bbd202b4b9312d812adcd7&end=cce71dc4adbb716bd87c9a5208f331d6cbfc7382&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | 2.6%     | [1.5%, 4.1%] | 4         |
| Regressions ❌ <br /> (secondary) | 2.7%     | [1.8%, 3.2%] | 13        |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -        | -     | 0         |
| All ❌✅ (primary) | 2.6%     | [1.5%, 4.1%] | 4         |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

Nilstrieb commented:
And they did magically go away and some even turned into improvements, very nice.

cjgillot commented:
Great. We will see about the `cache_on_disk` situation later.
@bors r+

bors commented:
:pushpin: Commit 4ae32c8cca22265c5c3ea8c2b751a251f53f97ed has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 4ae32c8cca22265c5c3ea8c2b751a251f53f97ed -->
<!-- homu: {"type":"Approved","sha":"4ae32c8cca22265c5c3ea8c2b751a251f53f97ed","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 4ae32c8cca22265c5c3ea8c2b751a251f53f97ed with merge f18001a215707d9635b0865ec00b420dc96c0d7a...
<!-- homu: {"type":"BuildStarted","head_sha":"4ae32c8cca22265c5c3ea8c2b751a251f53f97ed","merge_sha":"f18001a215707d9635b0865ec00b420dc96c0d7a"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3252535826/jobs/5339199013)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3252535826/jobs/5339199013","builder_name":"checks-actions"} -->

Nilstrieb commented:
I rebased the commits and tried running the rustdoc-gui tests locally (which for some reason fail here even though I touched nothing relevant to it??). Some tests failed locally but after a few retries they all passed. Are the rustdoc-gui tests that flaky? Anyways, this seems to be a flaky test and not the fault of this PR

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3252535826/jobs/5338856220) [(plain)](https://github.com/rust-lang-ci/rust/commit/f18001a215707d9635b0865ec00b420dc96c0d7a/checks/8900211295/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.......... (70/79)
.........  (79/79)


/checkout/src/test/rustdoc-gui/basic-code.goml basic-code... FAILED
[ERROR] (line 3) expected 1 elements, found 0: for command `assert-count: (".src-line-numbers", 1)`
Build completed unsuccessfully in 0:02:28
```

</details>

jyn514 commented:
@bors retry

Dylan-DPC commented:
@bors r=cjgillot

bors commented:
:pushpin: Commit 24ce4cfa205da0afc8500684465bf8158bd4adae has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 24ce4cfa205da0afc8500684465bf8158bd4adae -->
<!-- homu: {"type":"Approved","sha":"24ce4cfa205da0afc8500684465bf8158bd4adae","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 24ce4cfa205da0afc8500684465bf8158bd4adae with merge b8c35ca26b191bb9a9ac669a4b3f4d3d52d97fb1...
<!-- homu: {"type":"BuildStarted","head_sha":"24ce4cfa205da0afc8500684465bf8158bd4adae","merge_sha":"b8c35ca26b191bb9a9ac669a4b3f4d3d52d97fb1"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3255860823/jobs/5346664816)
Approved by: cjgillot
Pushing b8c35ca26b191bb9a9ac669a4b3f4d3d52d97fb1 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cjgillot","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3255860823/jobs/5346664816"},"merge_sha":"b8c35ca26b191bb9a9ac669a4b3f4d3d52d97fb1"} -->

rust-timer commented:
Finished benchmarking commit (b8c35ca26b191bb9a9ac669a4b3f4d3d52d97fb1): [comparison URL](https://perf.rust-lang.org/compare.html?start=c93ef33700e4e4f84fd85690df71b14c1d2b0aa3&end=b8c35ca26b191bb9a9ac669a4b3f4d3d52d97fb1&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=c93ef33700e4e4f84fd85690df71b14c1d2b0aa3&end=b8c35ca26b191bb9a9ac669a4b3f4d3d52d97fb1&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | 0.3%     | [0.2%, 0.5%] | 9         |
| Regressions ❌ <br /> (secondary) | 0.6%     | [0.3%, 0.8%] | 5         |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -0.7%    | [-1.2%, -0.2%] | 6         |
| All ❌✅ (primary) | 0.3%     | [0.2%, 0.5%] | 9         |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=c93ef33700e4e4f84fd85690df71b14c1d2b0aa3&end=b8c35ca26b191bb9a9ac669a4b3f4d3d52d97fb1&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | 2.0%     | [2.0%, 2.0%] | 1         |
| Regressions ❌ <br /> (secondary) | -        | -     | 0         |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -2.9%    | [-3.7%, -2.1%] | 9         |
| All ❌✅ (primary) | 2.0%     | [2.0%, 2.0%] | 1         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=c93ef33700e4e4f84fd85690df71b14c1d2b0aa3&end=b8c35ca26b191bb9a9ac669a4b3f4d3d52d97fb1&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | -        | -     | 0         |
| Regressions ❌ <br /> (secondary) | -        | -     | 0         |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -2.7%    | [-3.4%, -2.1%] | 13        |
| All ❌✅ (primary) | -        | -     | 0         |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*
